## STRATEGY for getting AMR gene lists

Even if we narrow our search on only ESKAPE pathogens, the number of AMR genes is still pretty high ~ 3.5million entries for ~ 250,000 unique genomes, so here's my plan of action :
Extract the AMR genes for each pathogen entry by filtering on the following:
- subtype = 'AMR'
- scope = 'core'
- select first 15 unique genomes


The above steps done for 6 pathogen species:
- Enterococcus faecium
- Staphylococcus aureus
- Klebsiella pneumoniae
- Acinetobacter baumannii
- Pseudomonas aeruginosa
- other Enterobacterales group

## NEW strategy!!

(by Madeline and Parul)

1. Downloaded list of all ESKAPE pathogens from ISOLATE dataset wtih:
   - LEVEL = 'chromosome'
   - LEVEL = 'complete'
2. Download all the 'protein-fasta' files using the Genome Accession Ids through ncbi-genome-download tool
3. Download list of all AMR genes belonging to ESKAPE pathogens from microbiggE dataset with:
   - subtype = 'AMR'
   - scope = 'core'
4. Match the AMR genes (ID) with the genes (ID) in the protein-fasta to tag them as AMR v/s non-AMR.



In [4]:
import pandas as pd

In [ ]:
#read and combine data from tsv files 
#df1 = pd.read_csv('AMR_metadata/E.tsv', sep='\t')
#df2 = pd.read_csv('AMR_metadata/S.tsv', sep='\t')
#df3 = pd.read_csv('AMR_metadata/K.tsv', sep='\t')
#df4 = pd.read_csv('AMR_metadata/A.tsv', sep='\t')
#df5 = pd.read_csv('AMR_metadata/P.tsv', sep='\t')
#df6 = pd.read_csv('AMR_metadata/otherE.tsv', sep='\t')

#combined_AMR_metadata = pd.concat([df1,df2,df3,df4,df5,df6], ignore_index= 'TRUE')
#combined_AMR_metadata

### Now let's get the list of genome assemblies we have to download

In [85]:
# assemblies = combined_AMR_metadata['Assembly'].unique().tolist()
# print (assemblies)

# # Save the list to a file
# with open('AMR_metadata/unique_assemblies.txt', 'w') as f:
#     for element in assemblies:
#         f.write(f"{element}\n")

In [86]:
# ! head -10 AMR_metadata/unique_assemblies.txt

### Downloading the above list of genomes using ncbi-genome-download

In [25]:
conda install bioconda::ncbi-genome-download


Channels:
 - defaults
 - bioconda
Platform: linux-64
Solving environment: done

## Package Plan ##

  environment location: /home/maaly7/anaconda3/envs/general

  added / updated specs:
    - bioconda::ncbi-genome-download


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    brotli-python-1.0.9        |  py311h6a678d5_8         358 KB
    certifi-2024.8.30          |  py311h06a4308_0         163 KB
    charset-normalizer-3.3.2   |     pyhd3eb1b0_0          44 KB
    idna-3.7                   |  py311h06a4308_0         133 KB
    ncbi-genome-download-0.3.3 |     pyh7cba7a3_0          29 KB  bioconda
    pysocks-1.7.1              |  py311h06a4308_0          35 KB
    requests-2.32.3            |  py311h06a4308_0         126 KB
    tqdm-4.66.5                |  py311h92b7b1e_0         163 KB
    urllib3-2.2.2              |  py311h06a4308_0         227 KB
    --------------------------------

In [29]:
! ncbi-genome-download --help

usage: ncbi-genome-download [-h] [-s {refseq,genbank}] [-F FILE_FORMATS]
                            [-l ASSEMBLY_LEVELS] [-g GENERA] [--genus GENERA]
                            [--fuzzy-genus] [-S STRAINS] [-T SPECIES_TAXIDS]
                            [-t TAXIDS] [-A ASSEMBLY_ACCESSIONS]
                            [--fuzzy-accessions] [-R REFSEQ_CATEGORIES]
                            [--refseq-category REFSEQ_CATEGORIES] [-o OUTPUT]
                            [--flat-output] [-H] [-P] [-u URI] [-p N] [-r N]
                            [-m METADATA_TABLE] [-n] [-N] [-v] [-d] [-V]
                            [-M TYPE_MATERIALS]
                            groups

positional arguments:
  groups                The NCBI taxonomic groups to download (default: all).
                        A comma-separated list of taxonomic groups is also
                        possible. For example: "bacteria,viral"Choose from:
                        ['all', 'archaea', 'bacteria', 'fungi',
        

### List of genomes to download (directly downloaded from Isolates browser) are the following:
GCAs_to_download_Sa.txt \
GCAs_to_download_Pa.txt \
GCAs_to_download_Kp.txt \
GCAs_to_download_Ef.txt \
GCAs_to_download_E.txt \
GCAs_to_download_Ab.txt

In [50]:
 ! ncbi-genome-download -s genbank -F 'cds-fasta' -A 'GCA_000009645.1' -o genome_assemblies/ --flat-output -v bacteria

INFO: Using cached summary.
INFO: Checking record 'GCA_000009645.1'


In [ ]:
! ncbi-genome-download -s genbank -F 'translated-cds' -A 'GCA_000009645.1' -o genome_assemblies/ --flat-output -v bacteria

In [51]:
! gzip -d genome_assemblies/GCA_000009645.1_ASM964v1_cds_from_genomic.fna.gz

In [53]:
! head -20 genome_assemblies/GCA_000009645.1_ASM964v1_cds_from_genomic.fna

>lcl|BA000018.3_cds_BAB41217.1_1 [gene=dnaA] [protein=chromosomal replication initiator protein] [protein_id=BAB41217.1] [location=517..1878] [gbkey=CDS]
ATGTCGGAAAAAGAAATTTGGGAAAAAGTGCTTGAAATTGCTCAAGAAAAATTATCAGCTGTAAGTTACTCAACTTTCCT
AAAAGATACTGAGCTTTACACGATTAAAGATGGTGAAGCTATCGTATTATCGAGTATTCCTTTTAATGCAAATTGGTTAA
ATCAACAATATGCTGAAATTATCCAAGCAATCTTATTTGATGTTGTAGGCTATGAAGTTAAACCTCACTTTATTACTACT
GAAGAATTAGCAAATTATAGTAATAATGAAACTGCTACTCCAAAAGAAACAACAAAACCTTCTACTGAAACAACTGAGGA
TAATCATGTGCTTGGTAGAGAGCAATTCAATGCCCATAACACATTTGACACTTTTGTAATCGGACCTGGTAACCGCTTTC
CACATGCAGCAAGTTTAGCTGTAGCCGAAGCACCAGCCAAAGCGTACAATCCATTATTTATCTATGGAGGGGTTGGTTTA
GGAAAAACCCATTTAATGCATGCCATTGGTCATCATGTTTTAGATAATAATCCAGATGCCAAAGTGATTTACACATCAAG
TGAAAAATTCACAAATGAATTTATTAAATCAATTCGTGATAACGAAGGTGAAGCTTTCAGAGAAAGATATCGTAATATCG
ACGTCTTATTAATCGATGATATTCAGTTCATACAAAATAAAGTACAAACACAAGAAGAATTTTTCTATACTTTTAATGAA
TTGCATCAGAATAACAAGCAAATAGTTATTTCGAGTGATCGACCACCAAAGGAAATTGCACAATTAGAAGACCGATTACG
TTCACGCTTTGAATGGGGGCTAATTGTTGATATTAC

In [48]:
! grep 'WP_001242578.1' genome_assemblies/GCA_000009645.1_ASM964v1_translated_cds.faa

In [82]:
#Don't run this again!!!!
! ncbi-genome-download -s genbank -F 'protein-fasta' -A GCAs_to_download_Pa.txt -o protein_fastas/ --flat-output -v bacteria

INFO: Using cached summary.
INFO: Checking record 'GCA_000750905.1'
INFO: Checking record 'GCA_000763245.3'
INFO: Checking record 'GCA_000816985.1'
INFO: Checking record 'GCA_000829255.1'
INFO: Checking record 'GCA_000829275.1'
INFO: Checking record 'GCA_000829885.1'
INFO: Checking record 'GCA_000981825.1'
INFO: Checking record 'GCA_000988505.1'
INFO: Checking record 'GCA_001077475.1'
INFO: Checking record 'GCA_001293085.1'
INFO: Checking record 'GCA_001447845.1'
INFO: Checking record 'GCA_001457615.1'
INFO: Checking record 'GCA_001482325.1'
INFO: Checking record 'GCA_001515845.2'
INFO: Checking record 'GCA_001515915.2'
INFO: Checking record 'GCA_001516005.1'
INFO: Checking record 'GCA_001516105.1'
INFO: Checking record 'GCA_001516165.2'
INFO: Checking record 'GCA_001516185.2'
INFO: Checking record 'GCA_001516205.2'
INFO: Checking record 'GCA_001516225.2'
INFO: Checking record 'GCA_001516245.2'
INFO: Checking record 'GCA_001516265.1'
INFO: Checking record 'GCA_001516305.2'
INFO: Checki

In [5]:
! ls protein_fastas | wc -l

8267


In [6]:
! ls protein_fastas | head -20

GCA_000006765.1_ASM676v1_protein.faa
GCA_000009005.1_ASM900v1_protein.faa
GCA_000009585.1_ASM958v1_protein.faa
GCA_000009645.1_ASM964v1_protein.faa
GCA_000009665.1_ASM966v1_protein.faa
GCA_000009885.1_ASM988v1_protein.faa
GCA_000010445.1_ASM1044v1_protein.faa
GCA_000010465.1_ASM1046v1_protein.faa
GCA_000011265.1_ASM1126v1_protein.faa
GCA_000011505.1_ASM1150v1_protein.faa
GCA_000011525.1_ASM1152v1_protein.faa
GCA_000012045.1_ASM1204v1_protein.faa
GCA_000013425.1_ASM1342v1_protein.faa
GCA_000013465.1_ASM1346v1_protein.faa
GCA_000014625.1_ASM1462v1_protein.faa
GCA_000016305.1_ASM1630v1_protein.faa
GCA_000016805.1_ASM1680v1_protein.faa
GCA_000017085.1_ASM1708v1_protein.faa
GCA_000017205.1_ASM1720v1_protein.faa
GCA_000018445.1_ASM1844v1_protein.faa
ls: write error: Broken pipe


In [73]:
! head -5 GCAs_to_download_Pa.txt > temp.txt
! head -5 GCAs_to_download_Sa.txt >> temp.txt
! head -5 GCAs_to_download_Kp.txt >> temp.txt
! head -5 GCAs_to_download_Ef.txt >> temp.txt
! head -5 GCAs_to_download_E.txt >> temp.txt
! head -5 GCAs_to_download_Ab.txt >> temp.txt

In [74]:
! head -40 temp.txt

GCA_000006765.1
GCA_000014625.1
GCA_000017205.1
GCA_000026645.1
GCA_000168335.1
GCA_000009005.1
GCA_000009645.1
GCA_000009665.1
GCA_000010445.1
GCA_000010465.1
GCA_000009885.1
GCA_000016305.1
GCA_000019565.1
GCA_000025465.1
GCA_000215745.1
GCA_000174395.2
GCA_000336405.1
GCA_000411655.2
GCA_000737555.1
GCA_001298485.1
GCA_000025565.1
GCA_000235765.3
GCA_000724505.1
GCA_000770155.1
GCA_001708345.1
GCA_000018445.1
GCA_000021245.2
GCA_000069245.1
GCA_000196795.1
GCA_000191145.1


In [87]:
! cat temp.txt | xargs -I {} find protein_fastas/ -name "{}*.faa" -print > temp2.txt

In [88]:
! xargs -a temp2.txt -I {} mv {} sub-samples/

###  This means we have 8,267 unique genomes (with protein fastas available) from the ESKAPE pathogens  


In [91]:
#Extract all the gunzipped files  ------ Don't run this again
! gzip -d protein_fastas/*.gz

In [7]:
#! head -20 genome_assemblies/GCA_018086425.1_PDT000701543.1_cds_from_genomic.fna
! head -20 protein_fastas/GCA_000006765.1_ASM676v1_protein.faa

>AAG03391.1 chromosomal replication initiator protein DnaA [Pseudomonas aeruginosa PAO1]
MSVELWQQCVDLLRDELPSQQFNTWIRPLQVEAEGDELRVYAPNRFVLDWVNEKYLGRLLELLGERGEGQLPALSLLIGS
KRSRTPRAAIVPSQTHVAPPPPVAPPPAPVQPVSAAPVVVPREELPPVTTAPSVSSDPYEPEEPSIDPLAAAMPAGAAPA
VRTERNVQVEGALKHTSYLNRTFTFENFVEGKSNQLARAAAWQVADNLKHGYNPLFLYGGVGLGKTHLMHAVGNHLLKKN
PNAKVVYLHSERFVADMVKALQLNAINEFKRFYRSVDALLIDDIQFFARKERSQEEFFHTFNALLEGGQQVILTSDRYPK
EIEGLEERLKSRFGWGLTVAVEPPELETRVAILMKKAEQAKIELPHDAAFFIAQRIRSNVRELEGALKRVIAHSHFMGRP
ITIELIRESLKDLLALQDKLVSIDNIQRTVAEYYKIKISDLLSKRRSRSVARPRQVAMALSKELTNHSLPEIGVAFGGRD
HTTVLHACRKIAQLRESDADIREDYKNLLRTLTT
>AAG03392.1 DNA polymerase III, beta chain [Pseudomonas aeruginosa PAO1]
MHFTIQREALLKPLQLVAGVVERRQTLPVLSNVLLVVEGQQLSLTGTDLEVELVGRVVLEDAAEPGEITVPARKLMDICK
SLPNDVLIDIRVEEQKLLVKAGRSRFTLSTLPANDFPTVEEGPGSLNFSIAQSKLRRLIDRTSFAMAQQDVRYYLNGMLL
EVNGGTLRSVATDGHRLAMCSLDAQIPSQDRHQVIVPRKGILELARLLTEQDGEVGIVLGQHHIRATTGEFTFTSKLVDG
KFPDYERVLPRGGDKLVVGDRQQLREAFSRTAILSNEKYRGIRLQLSNGLLKIQANNPEQEEAEEEVQVEYNGG

## Converting the fasta files into 1 big combined table

In [8]:
from Bio import SeqIO
import os

In [9]:
def create_cds_dataframe(fasta_dir):
    """Creates a DataFrame from all cds FASTA files in a directory."""
    
    data = []
    
    # Iterate over all files in the directory
    for filename in os.listdir(fasta_dir):
        # Check if the file ends with '.faa'
        if filename.endswith(".faa"):
            filepath = os.path.join(fasta_dir, filename)
            
            # Parse the FASTA file and extract headers and sequences
            for record in SeqIO.parse(filepath, "fasta"):
                data.append({"header": record.id, "sequence": str(record.seq)})
    
    return pd.DataFrame(data)


In [89]:
fasta_dir = "sub-samples/"  

# Create the DataFrame from the FASTA files
sub_sample_all_df = create_cds_dataframe(fasta_dir)

print(sub_sample_all_df)

            header                                           sequence
0       CAI79689.1  MSEKEIWEKVLKIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSSI...
1       CAI79690.1  MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVILT...
2       CAI79691.1  MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLINGV...
3       CAI79692.1  MKLNTLQLENYRNYDEVTLKCHPDVNILIGENAQGKTNLLESIYTL...
4       CAI79693.1  MVTALSDVNNTDNYGAGQIQVLEGLEAVRKRPGMYIGSTSERGLHH...
...            ...                                                ...
118374  UDP42675.1  MNLFKEEIGNWQDWSNVSQSLFAFSALTNFIFQKENLPLAKLENLP...
118375  UDP42676.1  MDYGFISTIVRSELFMMQLDSVLVSGAQPNVLSKEIDSFNFMIPIL...
118376  UDP42677.1  MPIAIGNKRLPVTLDEKRQKELQQLKQKYGKSESRIMCIALDLLIA...
118377  UDP42678.1  DPVSFKPIYGRSGKLETAYWGSRASERQVRMYNKKLEQERKKQIVP...
118378  UDP42679.1  MANENQLSEIQSKIEGRGAEKRTTINQLEKLNRISNRRLSVDDKVH...

[118379 rows x 2 columns]


In [11]:
#all_df['header'][1]  #### the header is the protein ID !!

'QQF40241.1'

In [12]:
# all_df.to_csv("all_protein_df.tsv", sep='\t', index=False)  ##Save into a df

In [15]:
#####Subsample the big df

# sampled_df = all_df.sample(n=10000, random_state=42)

# sampled_df.to_csv("subsampled_protein_df.tsv", sep='\t', index=False)  ##Save into a df

### Now match the protien IDs with the genes from the AMR table to get AMR v/s non-AMR protein sequences

In [102]:
# combined_AMR_metadata['Protein'].tolist()

# protein_ids = combined_AMR_metadata['Protein'].tolist()
# # Remove NaN values from the list
# protein_ids = [protein_id for protein_id in protein_ids if pd.notna(protein_id)]
# len(protein_ids)

In [16]:
microbigge_df = pd.read_csv('AMR_metadata/All.tsv', sep='\t')
microbigge_df

#Scientific name         Protein  \
0               Staphylococcus aureus subsp. aureus N315  WP_001242578.1   
1        Staphylococcus aureus subsp. aureus str. Newman  WP_001100300.1   
2                Staphylococcus aureus subsp. aureus MW2  WP_001801873.1   
3            Staphylococcus aureus subsp. aureus MSSA476  WP_001033157.1   
4                Staphylococcus aureus subsp. aureus COL  WP_063852655.1   
...                                                  ...             ...   
4050103                          Acinetobacter baumannii    HAV4520103.1   
4050104                          Acinetobacter baumannii    HAV4520104.1   
4050105                          Acinetobacter baumannii    HAV4520105.1   
4050106                          Acinetobacter baumannii    HAV4520106.1   
4050107                          Acinetobacter baumannii    HAV4520107.1   

                Assembly Scope Subtype  
0        GCA_000009645.1  core     AMR  
1        GCA_000010465.1  core     AMR  
2        GCA_000011265.1  core     AMR  
3        GCA_000011525.1  core     AMR  
4        GCA_000012045.1  core     AMR  
...                  ...   ...     ...  
4050103  GCA_016498365.1  core     AMR  
4050104  GCA_016498365.1  core     AMR  
4050105  GCA_016498365.1  core     AMR  
4050106  GCA_016498365.1  core     AMR  
4050107  GCA_016498365.1  core     AMR  

[4050108 rows x 5 columns]

#### Some summary stats:
- Total AMR genes found in the microbigge table for all ESKAPE pathogens = 4,050,108
- Of this list, AMR genes with Protein IDs listed = 3,451,878

In [102]:
#extract the same Genomes that I selecetd for the sub-samples
with open('temp.txt', 'r') as file:
    genome_ids = [genome_ids.strip() for genome_ids in file]
genome_ids

['GCA_000006765.1',
 'GCA_000014625.1',
 'GCA_000017205.1',
 'GCA_000026645.1',
 'GCA_000168335.1',
 'GCA_000009005.1',
 'GCA_000009645.1',
 'GCA_000009665.1',
 'GCA_000010445.1',
 'GCA_000010465.1',
 'GCA_000009885.1',
 'GCA_000016305.1',
 'GCA_000019565.1',
 'GCA_000025465.1',
 'GCA_000215745.1',
 'GCA_000174395.2',
 'GCA_000336405.1',
 'GCA_000411655.2',
 'GCA_000737555.1',
 'GCA_001298485.1',
 'GCA_000025565.1',
 'GCA_000235765.3',
 'GCA_000724505.1',
 'GCA_000770155.1',
 'GCA_001708345.1',
 'GCA_000018445.1',
 'GCA_000021245.2',
 'GCA_000069245.1',
 'GCA_000196795.1',
 'GCA_000191145.1']

In [104]:
##select AMR genes in only the genomes listed in 'genome_ids'
genome_ids = set(genome_ids)
microbigge_subsample_df = microbigge_df[microbigge_df['Assembly'].isin(genome_ids)]

microbigge_subsample_df

#Scientific name         Protein  \
0               Staphylococcus aureus subsp. aureus N315  WP_001242578.1   
1        Staphylococcus aureus subsp. aureus str. Newman  WP_001100300.1   
7                          Acinetobacter baumannii ACICU  WP_002002480.1   
8                         Acinetobacter baumannii AB0057  WP_001159761.1   
55            Enterobacter cloacae subsp. dissolvens SDM  WP_043951555.1   
...                                                  ...             ...   
3927925                      Enterobacter cloacae ECNIH2  WP_000186237.1   
3949091                      Enterobacter cloacae ECNIH2  WP_000259031.1   
3980970                      Enterobacter cloacae ECNIH2  WP_012695470.1   
3987338                      Enterobacter cloacae ECNIH2  WP_004152396.1   
3993167                      Enterobacter cloacae ECNIH2  WP_000027057.1   

                Assembly Scope Subtype  
0        GCA_000009645.1  core     AMR  
1        GCA_000010465.1  core     AMR  
7        GCA_000018445.1  core     AMR  
8        GCA_000021245.2  core     AMR  
55       GCA_000235765.3  core     AMR  
...                  ...   ...     ...  
3927925  GCA_000724505.1  core     AMR  
3949091  GCA_000724505.1  core     AMR  
3980970  GCA_000724505.1  core     AMR  
3987338  GCA_000724505.1  core     AMR  
3993167  GCA_000724505.1  core     AMR  

[224 rows x 5 columns]

In [105]:
protein_ids = microbigge_subsample_df['Protein'].tolist()
# Remove NaN values from the list
protein_ids = [protein_id for protein_id in protein_ids if pd.notna(protein_id)]
len(protein_ids)

209

### THE PROTEIN ID in MICROBIGGE TABLE DOES NOT MATCH PROTEIN ACCESSION IN PROTEIN FASTA

### Hence downloading the protein sequences using protein IDS and match the sequences instead 

In [35]:
#! conda install bioconda::prodigal -y

In [39]:
#! pip install -r ../requirements.txt

In [42]:
# import requests as r
# from Bio import SeqIO
# from io import StringIO


In [ ]:
# baseUrl="http://www.uniprot.org/uniprot/"
# currentUrl=baseUrl+cID+".fasta"
# response = r.post(currentUrl)
# cData=''.join(response.text)

# Seq=StringIO(cData)
# pSeq=list(SeqIO.parse(Seq,'fasta'))

In [63]:
#### trying efetch 
! conda install bioconda::entrez-direct --yes

Channels:
 - defaults
 - bioconda
Platform: linux-64
Solving environment: done

## Package Plan ##

  environment location: /home/maaly7/anaconda3/envs/general

  added / updated specs:
    - bioconda::entrez-direct


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    entrez-direct-22.4         |       he881be0_0        14.0 MB  bioconda
    libidn2-2.3.4              |       h5eee18b_0         146 KB
    libunistring-0.9.10        |       h27cfd23_0         536 KB
    wget-1.24.5                |       h251f7ec_0         825 KB
    ------------------------------------------------------------
                                           Total:        15.5 MB

The following NEW packages will be INSTALLED:

  entrez-direct      bioconda/linux-64::entrez-direct-22.4-he881be0_0 
  libidn2            pkgs/main/linux-64::libidn2-2.3.4-h5eee18b_0 
  libunistring       pkgs/main/linux-64::libunistrin

In [66]:
! efetch --help

efetch 22.4

 NOTE:  EDirect version 22.6 is now available

Format Selection

  -format        Format of record or report
  -mode          text, xml, asn.1, json
  -style         master, conwithfeat

Direct Record Selection

  -db            Database name
  -id            Unique identifier or accession number
  -input         Read identifier(s) from file instead of stdin

Sequence Range

  -seq_start     First sequence position to retrieve
  -seq_stop      Last sequence position to retrieve
  -strand        1 = forward DNA strand, 2 = reverse complement
                   (otherwise strand minus is set if start > stop)
  -forward       Force strand 1
  -revcomp       Force strand 2

Gene Range

  -chr_start     Sequence range from 0-based coordinates
  -chr_stop        in gene docsum GenomicInfoType object

Sequence Flags

  -complexity    0 = default, 1 = bioseq, 3 = nuc-prot set
  -extend        Extend sequence retrieval in both directions
  -extrafeat     Bit flag specifying extra f

In [67]:
! efetch -db protein -format fasta -id WP_001242578.1

>WP_001242578.1 MULTISPECIES: bleomycin binding protein [Bacteria]
MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVRIHLWEASDEGWRSRSNDSPVCTGA
ESFIAGTASCRIEVEGIDELYQHIKPLGILHPNTSLKDQWWDERDFAVIDPDNNLISFFQQIKS


#### Running efetch on bulk to get all protein sequences in a dataframe format

In [106]:
import subprocess

# Initialize an empty list to store the data
protein_data = []

# Loop through each protein ID
for protein_id in protein_ids:
    # Run the efetch command using subprocess
    result = subprocess.run(
        ['efetch', '-db', 'protein', '-format', 'fasta', '-id', protein_id],
        capture_output=True, text=True
    )
    
    # Get the output from efetch (FASTA format)
    fasta_output = result.stdout

    # Extract the sequence (everything after the first line)
    fasta_lines = fasta_output.split('\n')
    sequence = ''.join(fasta_lines[1:])  # Join lines after the header to get the full sequence
    
    # Append protein ID and sequence to the data list
    protein_data.append({'protein_ID': protein_id, 'sequence': sequence})

# Convert the list to a DataFrame
df = pd.DataFrame(protein_data)

print(df)


         protein_ID                                           sequence
0    WP_001242578.1  MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVR...
1    WP_001100300.1  MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTT...
2    WP_002002480.1  MKLLKILSLVCLSISIGACAEHSMSRAKTSTIPQVNNSIIDQNVQA...
3    WP_001159761.1  MQFKKISCLLLSPLFIFSTSIYADNTPKDQEIKKLVDQNFKPLLEK...
4    WP_043951555.1  MMKKSLSCALLLSVACSAFAAPMSEKQLANVVERNVTPLMKAQAIP...
..              ...                                                ...
204  WP_000186237.1  MTNYFDSPFKGKLLSEQVKNPNIKVGRYSYYSGYYHGHSFDDCARY...
205  WP_000259031.1  MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGP...
206  WP_012695470.1  MSHIQRETSCSRPRLNSNLDVDLYGYRWARDNVGQSGATIYRLYGK...
207  WP_004152396.1  MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGV...
208  WP_000027057.1  MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIE...

[209 rows x 2 columns]


In [108]:
amr_protein_df = df
amr_protein_df

protein_ID                                           sequence
0    WP_001242578.1  MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVR...
1    WP_001100300.1  MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTT...
2    WP_002002480.1  MKLLKILSLVCLSISIGACAEHSMSRAKTSTIPQVNNSIIDQNVQA...
3    WP_001159761.1  MQFKKISCLLLSPLFIFSTSIYADNTPKDQEIKKLVDQNFKPLLEK...
4    WP_043951555.1  MMKKSLSCALLLSVACSAFAAPMSEKQLANVVERNVTPLMKAQAIP...
..              ...                                                ...
204  WP_000186237.1  MTNYFDSPFKGKLLSEQVKNPNIKVGRYSYYSGYYHGHSFDDCARY...
205  WP_000259031.1  MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGP...
206  WP_012695470.1  MSHIQRETSCSRPRLNSNLDVDLYGYRWARDNVGQSGATIYRLYGK...
207  WP_004152396.1  MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGV...
208  WP_000027057.1  MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIE...

[209 rows x 2 columns]

### Get the AMR v/s non-AMR sequences by matching the Protein sequence in both

In [1]:
# amr_df = pd.DataFrame()
# #protein_ids = combined_AMR_metadata['Protein'].tolist()

# # Iterate over the list of protein IDs and collect matching rows
# for protein_id in protein_ids:
#     # Get rows that match the protein ID in the 'header' column
#     matches = all_df[all_df['header'] == protein_id]
    
#     # Append matching rows to the matching_df
#     amr_df = pd.concat([amr_df, matches])

# # After processing all protein IDs, the non-matching rows are the remaining ones
# non_amr_df = all_df[~all_df.index.isin(amr_df.index)]


In [110]:
####tring a faster way instead!
# Convert the list of protein sequences to a set for faster lookup
sequence_set = set(amr_protein_df['sequence'])
sequence_set

{'MAKLLIMSVVSFCFIFLLLVFFRYILKRYFNYSLNYKVWYLTMLAGLIPFIPIKFSFIKFNNVNNQAPTVESKSHDLNHNINTTKPIQEFTTDIHKFNWDSIDNICTVIWIVLVIILSFKFLKALLYLKYLKKQSLYLNENEKNKIDTILFNHQYKKNIVIRKAEAIQSPITFWYGKYIILIPSSYFKSVIDKRLKYIILHEYAHAKNRDTLHLIIFNIFSIIMSYNPLVHIVKRKIIHDNEVEADRFVLNNINKNEFKTYAESIMDSVLKTPFSNKNILSHSFNGKKSLLKSRLINIKEADLKKQSKLILIFICIFTFFIMIIQSQFLMGQSLTDYNYKKPLQSDYQILDESKNFGSNSGSFVMYSMKKDKYYIYNEKESRKRYSPDSTYKIYLALFGLDRHIISDKNSRMSWNHKHYLFESWNKEQDLNTAMQNSVNWYFERISNQIPKNYTAAQLKQLNYGNENLGSYKSYWMEDSLKISNLEQVIVFKNMMEQNNHFSKKAKNQLSSSLLIKKNEKYELYGKTGTGIVNGKYNNGWFVGYVITNHDKYYFATHLSDGNPSGKNAELISEKILKGMGVLNDQ',
 'MAKMRISPELKKLIEKYRCVKDTEGMSPAKVYKLVGENENLYLKMTDSRYKGTTYDVEREKDMMLWLEGKLPVPKVLHFERHDGWSNLLMSEADGVLCSEEYEDEQSPEKIIELYAECIRLFHSIDISDCPYTNSLDSRLAELDYLLNNDLADVDCENWEEDTPFKDPRELYDFLKTEKPEEELVFSHGDLGDSNIFVKDGKVSGFIDLGRSGRADKWYDIAFCVRSIREDIGEEQYVELFFDLLGIKPDWEKIKYYILLDELF',
 'MDFSRFFIDRPIFAAVLSILIFITGLIAIPLLPVSEYPDVVPPSVQVRAEYPGANPKVIAETVATPLEEAINGVENMMYMKSVAGSDGVLVTTVTFRPGTDPDQAQVQVQNRVAQAEARLPEDVRRLGITTQKQSPTLT

In [111]:
# Filter rows on an exact match for any protein sequence
# Here, we use a list comprehension to find headers that exactly match any protein sequence from the set
sub_sample_all_df['AMR_status'] = sub_sample_all_df['sequence'].apply(lambda x: "AMR" if x in sequence_set else "non-AMR")

In [224]:
sub_sample_all_df[sub_sample_all_df['AMR_status'] == 'AMR']

header                                           sequence  \
3089    ABR81255.1  M       R       L       L       F       F     ...   
3095    ABR81261.1  M       T       D       Q       A       T     ...   
3367    ABR81550.1  M       S       Y       T       R       V     ...   
3706    ABR81912.1  M       I       E       Q       D       G     ...   
6733    ABR85124.1  M       R       H       A       T       I     ...   
...            ...                                                ...   
118356  UDP42657.1  M       R       S       E       K       E     ...   
118357  UDP42658.1  M       N       K       N       I       K     ...   
118361  UDP42662.1  M       N       T       S       Y       S     ...   
118363  UDP42664.1  M       R       S       E       K       E     ...   
118367  UDP42668.1  M       L       K       Q       K       E     ...   

       AMR_status  
3089          AMR  
3095          AMR  
3367          AMR  
3706          AMR  
6733          AMR  
...           ...  
118356        AMR  
118357        AMR  
118361        AMR  
118363        AMR  
118367        AMR  

[180 rows x 3 columns]

In [225]:
sub_sample_all_df

header                                           sequence  \
0       CAI79689.1  M       S       E       K       E       I     ...   
1       CAI79690.1  M       M       E       F       T       I     ...   
2       CAI79691.1  M       I       I       L       V       Q     ...   
3       CAI79692.1  M       K       L       N       T       L     ...   
4       CAI79693.1  M       V       T       A       L       S     ...   
...            ...                                                ...   
118374  UDP42675.1  M       N       L       F       K       E     ...   
118375  UDP42676.1  M       D       Y       G       F       I     ...   
118376  UDP42677.1  M       P       I       A       I       G     ...   
118377  UDP42678.1  D       P       V       S       F       K     ...   
118378  UDP42679.1  M       A       N       E       N       Q     ...   

       AMR_status  
0         non-AMR  
1         non-AMR  
2         non-AMR  
3         non-AMR  
4         non-AMR  
...           ...  
118374    non-AMR  
118375    non-AMR  
118376    non-AMR  
118377    non-AMR  
118378    non-AMR  

[118379 rows x 3 columns]

#### Total number of AMR protein sequences found = 180
#### Total number of non-AMR protein sequences found = 118199

In [115]:

# Save this as a training dataset to a TSV file
sub_sample_all_df.to_csv("training_set.tsv", sep='\t', index=False)


#### ##Now trying transformer

### Formatting the training set for as input into the transformer model

In [240]:
import re
###first introduce spaces between each protein sequence character
sub_sample_all_df['sequence'] = sub_sample_all_df['sequence'].apply(lambda x: re.sub(r'\s+','',x))
sub_sample_all_df['sequence'] = sub_sample_all_df['sequence'].apply(lambda x: ' '.join(list(x)))
sub_sample_all_df

header                                           sequence  \
0       CAI79689.1  M S E K E I W E K V L K I A Q E K L S A V S Y ...   
1       CAI79690.1  M M E F T I K R D Y F I T Q L N D T L K A I S ...   
2       CAI79691.1  M I I L V Q E V V V E G D I N L G Q F L K T E ...   
3       CAI79692.1  M K L N T L Q L E N Y R N Y D E V T L K C H P ...   
4       CAI79693.1  M V T A L S D V N N T D N Y G A G Q I Q V L E ...   
...            ...                                                ...   
118374  UDP42675.1  M N L F K E E I G N W Q D W S N V S Q S L F A ...   
118375  UDP42676.1  M D Y G F I S T I V R S E L F M M Q L D S V L ...   
118376  UDP42677.1  M P I A I G N K R L P V T L D E K R Q K E L Q ...   
118377  UDP42678.1  D P V S F K P I Y G R S G K L E T A Y W G S R ...   
118378  UDP42679.1  M A N E N Q L S E I Q S K I E G R G A E K R T ...   

       AMR_status  
0         non-AMR  
1         non-AMR  
2         non-AMR  
3         non-AMR  
4         non-AMR  
...           ...  
118374    non-AMR  
118375    non-AMR  
118376    non-AMR  
118377    non-AMR  
118378    non-AMR  

[118379 rows x 3 columns]

In [120]:
## split between training and test dataset

! pip install datasets
from datasets import Dataset, DatasetDict
import pandas as pd

In [241]:
amr_df = sub_sample_all_df[sub_sample_all_df['AMR_status']=='AMR']
non_amr_df = sub_sample_all_df[sub_sample_all_df['AMR_status']=='non-AMR'].sample(720,random_state=7)

In [242]:
hf_dataset = pd.concat([non_amr_df,amr_df]).reset_index(drop=True)
hf_dataset = Dataset.from_pandas(hf_dataset)
hf_dataset

Dataset({
    features: ['header', 'sequence', 'AMR_status'],
    num_rows: 900
})

In [243]:
# # Convert to Hugging Face Dataset
# hf_dataset = Dataset.from_pandas(sub_sample_all_df)
# hf_dataset

In [244]:
# Split into train and test datasets (80% train, 20% test)
train_test = hf_dataset.train_test_split(test_size=0.2)

In [245]:
# Split the train dataset further into train and validation sets (80% train, 20% validation)
train_valid = train_test['train'].train_test_split(test_size=0.2)

In [246]:
# Combine train, validation, and test sets into a single DatasetDict
final_splits = DatasetDict({
    'train': train_valid['train'],
    'validation': train_valid['test'],
    'test': train_test['test']
})

# Print the final splits
print(final_splits)

DatasetDict({
    train: Dataset({
        features: ['header', 'sequence', 'AMR_status'],
        num_rows: 576
    })
    validation: Dataset({
        features: ['header', 'sequence', 'AMR_status'],
        num_rows: 144
    })
    test: Dataset({
        features: ['header', 'sequence', 'AMR_status'],
        num_rows: 180
    })
})


In [247]:
final_splits['test'][0]

{'header': 'BAB42435.1',
 'sequence': 'M F N E K D Q L A V D T L R A L S I D T I E K A N S G H P G L P M G A A P M A Y T L W T R H L N F N P Q S K D Y F N R D R F V L S A G H G S A L L Y S L L H V S G S L E L E E L K Q F R Q W G S K T P G H P E Y R H T D G V E V T T G P L G Q G F A M S V G L A L A E D H L A G K F N K E G Y N V V D H Y T Y V L A S D G D L M E G I S H E A A S F A G H N K L S K L V V L Y D S N D I S L D G E L N K A F S E N T K A R F E A Y G W N Y L L V K D G N D L E E I D K A I T T A K S Q E G P T I I E V K T T I G F G S P N K A G T N G V H G A P L G E V E R K L T F E N Y G L D P E K R F N V S E E V Y E I F Q N T M L K R A N E D E S Q W N S L L E K Y A E T Y P E L A E E F K L A I S G K L P K N Y K D E L P R F E L G H N G A S R A D S G T V I Q A I S K T V P S F F G G S A D L A G S N K S N V N D A T D Y S S E T P E G K N V W F G V R E F A M G A A V N G M A A H G G L H P Y G A T F F V F S D Y L K P A L R L S S I M G L N A T F I F T H D S I A V G E D G P T H E P I E Q L A G L

#### Setting up transformer model to run with the above dataset

In [248]:
# ! pip install transformers datasets evaluate accelerate

In [262]:
from transformers import BertTokenizer, BertForSequenceClassification
tokenizer = BertTokenizer.from_pretrained('Rostlab/prot_bert')
model = BertForSequenceClassification.from_pretrained('Rostlab/prot_bert', num_labels=2)  # Assuming binary classification

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at Rostlab/prot_bert and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [263]:
from datasets import DatasetDict
from transformers import BertTokenizer

# Dataset should be in DatasetDict
def preprocess_function(examples):
    tokenized = tokenizer(examples["sequence"], padding="max_length", truncation=True, max_length=340)
    tokenized["labels"] = [1 if status == "AMR" else 0 for status in examples["AMR_status"]]    
    return tokenized


# Tokenize your dataset
tokenized_datasets = final_splits.map(preprocess_function, batched=True)


Map:   0%|          | 0/576 [00:00<?, ? examples/s]

Map:   0%|          | 0/144 [00:00<?, ? examples/s]

Map:   0%|          | 0/180 [00:00<?, ? examples/s]

In [264]:
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score

def compute_metrics(p):
    preds = p.predictions.argmax(-1)  # Get the predicted class (for binary classification, will give 0 or 1)
    labels = p.label_ids  # labels (0 or 1)
    precision = precision_score(labels, preds, average='binary')
    recall = recall_score(labels, preds, average='binary')
    f1 = f1_score(labels, preds, average='binary')
    accuracy = accuracy_score(labels, preds)

    return {
        'accuracy': accuracy,
        'precision': precision,
        'recall': recall,
        'f1': f1
    }


In [265]:
from transformers import Trainer, TrainingArguments

training_args = TrainingArguments(
    output_dir="./fine-tuned-prot-bert",
    evaluation_strategy="epoch",
    learning_rate=5e-5,
    per_device_train_batch_size=32,
    per_device_eval_batch_size=32,
    num_train_epochs=10,
    weight_decay=0.01,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets['train'],
    eval_dataset=tokenized_datasets['validation'],
    compute_metrics=compute_metrics
)

trainer.train()


/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/transformers/training_args.py:1545: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '


/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/maaly7/anaconda3/envs/gener

TrainOutput(global_step=50, training_loss=0.22764545440673828, metrics={'train_runtime': 100.3408, 'train_samples_per_second': 57.404, 'train_steps_per_second': 0.498, 'total_flos': 4452701487667200.0, 'train_loss': 0.22764545440673828, 'epoch': 10.0})

In [253]:
results = trainer.evaluate(tokenized_datasets['test'])
print(results)

/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '


{'eval_loss': 0.053683504462242126, 'eval_accuracy': 0.9888888888888889, 'eval_precision': 1.0, 'eval_recall': 0.9285714285714286, 'eval_f1': 0.9629629629629629, 'eval_runtime': 0.897, 'eval_samples_per_second': 200.662, 'eval_steps_per_second': 2.23, 'epoch': 10.0}


In [276]:
import torch

# Check if a GPU is available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")


model.to(device)

sequence = "MTADERDRARSALPFLVITQLMIVLDASIVNIALPSMGRELGMDQTGLQWVVNAYTLTFGGFLMLGGRMADLIGRRLVFVSGICLFGAASLAAALAPVAGVLVAARAVQGLSAAVASAAALSIIVATFPEGKGRNQALAMWGAVSGVGGAVGVLLGGVLTSGPGWPWIFYINVPIVVVVVLGVFRSVSGARGDTRGRLDVAGAVTLTGGLTLLVYAIVSGQSGDPVTILLALGLAVVLLVSFFLVQRKVREPLVPLSSFRNRNLSVASVVGLFAGAAPYAMFFLLSLHLQNVVGLTPLQTGLGFLPVSLISMVGAAALAPLAMARIGMRFTLLLSLGVLAVGLVLLSRLTEEDGFGATVAGQLVAGLGLGTTFVAVTTAAVAGLAENESGLASGLINTAQQLGGALGLGALAALSGAYSAAELAKEPPVSEVAALSSGYQVAFLGAAVFAVAGALIALALPRRESVPATTPHE" 
# sequence = "MSVHIEPIGRFLLAVGVIVAVCHLGGLLCHRIRQPPVIGEIAAGLLLGPTLLGAVAPSLQRALFPEEVLQAVGMAAQLGLVTFMFLLGSELRVDHVRGNGKVVWALVAGSILLPFLAGTGFALLTRPAFGTPQVSTTAYALFVGLAMSITALPVLARILADFRADQSFLGTLALMAAAVGDALAWAALTVILAVTGSGSTGELVLRSALALTLVLLTVFVVKPALRTLLHRLPVNSRVTVPALVVGTTAFAATTEVIGLHPVIGAFLFGCAMPRGSAVLQRASAQLRGFTVSVLLPLFFAGVAMKTAFDAFGTAGNWLLFAAALAVATVTKFVGASSGALLAGLDRARAFQLGALMNCRGVTELVVATVGLQNGFVNEFGYTVLVLIALVTTALTGPLARLRAEEAPQENHRIPMKHGGTFHVRQD"
sequence = re.sub(r'\s+','',sequence)
sequence = ' '.join(list(sequence))

inputs = tokenizer(sequence, padding="max_length", truncation=True, max_length=340, return_tensors="pt")
inputs = {key: value.to(device) for key, value in inputs.items()}  # Move inputs to the same device


model.eval()

# Run inference with no_grad() to avoid gradient computations
with torch.no_grad():
    outputs = model(**inputs)

# Convert logits to probabilities using sigmoid (for binary classification)
probabilities = torch.sigmoid(outputs.logits)

# predictions is likely a 2-element tensor, so you need to get the index of the max value or threshold
predictions = (probabilities > 0.5).int()

# If it's a single sequence, you can convert the predictions tensor into a Python list
# Since this is binary classification, you likely want the first prediction
predicted_label = label_map[predictions.argmax().item()]  # Get the predicted class based on the index

print(f"The predicted label for the sequence is: {predicted_label}")


The predicted label for the sequence is: AMR
